### ISSUE TITLE
[VPA] Weird restarts ("multiple configs")

### ISSUE THREAD

**Issue by R-Studio** [kubernetes/autoscaler#6010]
**Which component are you using?**: vertical-pod-autoscaler (recommender, updater & admission-controller)

**What version of the component are you using?**: v0.13 (Image-Tag: 0.13.0, [Fairwirds-Helm Chart](https://artifacthub.io/packages/helm/fairwinds-stable/vpa): v1.7.2)

Component version:
**What k8s version are you using (`kubectl version`)?**: OpenShift 4.11, Kubernetes v1.24.12+ceaf338

**What environment is this in?**: OnPrem VMs

**What behaviour did you expect to see?:** No unreasonable pod evictions/restarts

**What happened instead?**: Unseasonable pod evictions/restarts
![image](https://github.com/kubernetes/autoscaler/assets/7498854/041beda8-e68b-432a-b2c9-3d16181d90d3)
In the following screenshot we can see that VPA set the CPU requests from 0.055 to 0.043 for 2 minutes and then back to 0.055 again:
![image](https://github.com/kubernetes/autoscaler/assets/7498854/0a0deb38-c8a3-449d-b3f4-29bf8b3d2ba1)
I have noticed the following log, **why 5 configs**? (I only have one `VerticalPodAutoscaler` for this deployment deployed)
`matcher.go:68] Let's choose from 5 configs for pod NAMESPACE/xxx-85b97d6dfc-%`

**How to reproduce it (as minimally and precisely as possible)**: I am not sure if it is reproducible

**I use the following arguments:**
Recommender:
```
v: "4"
target-cpu-percentile: 0.50
pod-recommendation-min-cpu-millicores: 10
pod-recommendation-min-memory-mb: 10
recommendation-margin-fraction: 0.0
memory-saver: true
```

Updater:
```
v: "4"
min-replicas: 1
```

AdmissionController: no arguments than the defaults

**Comment by R-Studio** [comment]
I noticed that vpa-updater protects the newly created pod only for 1min, why?
![image](https://github.com/kubernetes/autoscaler/assets/7498854/df93a6c1-a2dc-49b8-8931-6d123e1e243a)

**Comment by voelzmo** [comment]
Hey @R-Studio, thanks for providing some insight into your investigations! I'm not able to answer what's going on completely, so here's just a few pointers to clear the fog step-by-step:
* the log statement `Let's choose from X configs` is coming from the `admission-webhook` that reacts on CREATE Pod. [For each Pod, it is trying to find a matching VPA](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/admission-controller/resource/pod/handler.go#L80), and you currently have 5 VPAs in the Pod's namespace (`kubectl -n NAMESPACE get vpa` should show this)
* Looking at the single graph you showed makes it hard to follow what happened. My recommendation is to print additional information into the same diagram, [as the `updater` evicts a Pod in the following scenarios](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go#L118-L121):
  * the request is outside the recommended range for at least one container
  * the pod lives for at least 12h and the resource diff is >= MinChangePriority
  * a vpa scaled container OOMed in less than evictAfterOOMThreshold

In you case, we can probably rule out cases 2 and 3, given that you're only scaling on CPU (you didn't explicitly mention this, but I was assuming it from the way you investigated this) and that the Pod was evicted just minutes before. So most likely the current requests are outside the recommended range for one of the containers in your Pod. To analyze this, it helps to also draw the upper and lower bounds and see when/how a Container's current requests end up being outside the bounds of the new recommendation (which is case 1 above). This would look like this
<img width="2473" alt="Screenshot 2023-08-07 at 11 32 32" src="https://github.com/kubernetes/autoscaler/assets/2256887/80ce6c1e-516b-4cc5-a01b-eded390a95eb">


In this example graph, we see the "current requests" between the "lower bound", and "upper bound", so we're not in case 1. The new recommendation is lower than the current requests, though, and after 12 hours the Pod would be evicted to apply the new recommendation, as there is more than 10% difference.

A graph like this should show you, why your Pod is getting evicted and hopefully explain what happens.

**Comment by R-Studio** [comment]
@voelzmo thanks for your reply! 👍🏽 
First you are right we are only scaling on CPU requests (sorry I forgot to mention this).   
Here an example of our VPA resources:
```yaml
---
apiVersion: "autoscaling.k8s.io/v1"
kind: VerticalPodAutoscaler
metadata:
  name: vpa-argocd-applicationset-controller
  namespace: argocd
spec:
  targetRef:
    apiVersion: "apps/v1"
    kind: Deployment
    name: argocd-applicationset-controller
  updatePolicy:
    updateMode: "Auto"
  resourcePolicy:
    containerPolicies:
      - containerName: '*'
        minAllowed:
          cpu: 10m
        maxAllowed:
          cpu: 3
        controlledResources: ["cpu"]
        controlledValues: "RequestsOnly"
```

- Regarding to `Let's choose from X configs`: you are absolutely right. I thought vpa finds multiple configurations for the same pod 😄😅 
- Regarding to more information's in the graph: Unfortunately we are using OpenShift and there is a bug that does not allow us to collect the VPA metrics: https://access.redhat.com/solutions/6301491 

Anyway thanks for all your inputs! 👍🏽
